File: NuvioMobile/composeApp/src/commonMain/kotlin/com/nuvio/app/features/addons/AddonManifestParser.kt
package com.nuvio.app.features.addons

import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import nuvio.composeapp.generated.resources.Res
import nuvio.composeapp.generated.resources.addons_manifest_missing_field
import org.jetbrains.compose.resources.getString

internal object AddonManifestParser {
    private val json = Json {
        ignoreUnknownKeys = true
    }

    fun parse(
        manifestUrl: String,
        payload: String,
    ): AddonManifest {
        val root = json.parseToJsonElement(payload).jsonObject
        val defaultTypes = root.stringList("types")
        val defaultPrefixes = root.stringList("idPrefixes")

        return AddonManifest(
            id = root.requiredString("id"),
            name = root.requiredString("name"),
            description = root.optionalString("description").orEmpty(),
            version = root.requiredString("version"),
            logoUrl = root.optionalString("logo")?.resolveAgainstManifest(manifestUrl),
            resources = root.resources(defaultTypes, defaultPrefixes),
            types = defaultTypes,
            idPrefixes = defaultPrefixes,
            catalogs = root.catalogs(),
            behaviorHints = root.behaviorHints(),
            transportUrl = manifestUrl,
        )
    }

    private fun JsonObject.resources(
        defaultTypes: List<String>,
        defaultPrefixes: List<String>,
    ): List<AddonResource> =
        array("resources").map { resource ->
            when (resource) {
                is JsonPrimitive -> AddonResource(
                    name = resource.contentOrNull.orEmpty(),
                    types = defaultTypes,
                    idPrefixes = defaultPrefixes,
                )

                else -> {
                    val obj = resource.jsonObject
                    AddonResource(
                        name = obj.requiredString("name"),
                        types = obj.stringList("types").ifEmpty { defaultTypes },
                        idPrefixes = obj.stringList("idPrefixes").ifEmpty { defaultPrefixes },
                    )
                }
            }
        }.filter { it.name.isNotBlank() }

    private fun JsonObject.catalogs(): List<AddonCatalog> =
        array("catalogs").map { catalogElement ->
            val catalog = catalogElement.jsonObject
            AddonCatalog(
                type = catalog.requiredString("type"),
                id = catalog.requiredString("id"),
                name = catalog.optionalString("name").orEmpty().ifBlank { catalog.requiredString("id") },
                extra = catalog.array("extra").mapNotNull { extraElement ->
                    extraElement.jsonObject.optionalString("name")?.takeIf { it.isNotBlank() }?.let { name ->
                        AddonExtraProperty(
                            name = name,
                            isRequired = extraElement.jsonObject.boolean("isRequired"),
                            options = extraElement.jsonObject.stringList("options"),
                            optionsLimit = extraElement.jsonObject.int("optionsLimit"),
                        )
                    }
                },
            )
        }

    private fun JsonObject.behaviorHints(): AddonBehaviorHints {
        val hints = this["behaviorHints"]?.jsonObject ?: return AddonBehaviorHints()
        return AddonBehaviorHints(
            configurable = hints.boolean("configurable"),
            configurationRequired = hints.boolean("configurationRequired"),
            adult = hints.boolean("adult"),
            p2p = hints.boolean("p2p"),
        )
    }

    private fun JsonObject.requiredString(name: String): String =
        optionalString(name)?.takeIf { it.isNotBlank() }
            ?: throw IllegalArgumentException(
                runBlocking { getString(Res.string.addons_manifest_missing_field, name) },
            )

    private fun JsonObject.optionalString(name: String): String? =
        this[name]?.jsonPrimitive?.contentOrNull

    private fun JsonObject.stringList(name: String): List<String> =
        array(name).mapNotNull { element ->
            element.jsonPrimitive.contentOrNull?.takeIf { it.isNotBlank() }
        }

    private fun JsonObject.array(name: String): JsonArray =
        (this[name] as? JsonArray) ?: JsonArray(emptyList())

    private fun JsonObject.boolean(name: String): Boolean =
        this[name]?.jsonPrimitive?.booleanOrNull == true

    private fun JsonObject.int(name: String): Int? =
        this[name]?.jsonPrimitive?.contentOrNull?.toIntOrNull()
}

private fun String.resolveAgainstManifest(manifestUrl: String): String =
    when {
        startsWith("http://") || startsWith("https://") || startsWith("data:") -> this
        startsWith("//") -> "https:$this"
        else -> {
            val manifestBase = manifestUrl.substringBefore("?").substringBeforeLast('/', "")
            if (startsWith('/')) {
                val origin = manifestBase.substringBefore("/", missingDelimiterValue = manifestBase)
                val schemeAndHost = if (origin.contains("://")) {
                    val scheme = origin.substringBefore("://")
                    val host = manifestBase.substringAfter("://").substringBefore("/")
                    "$scheme://$host"
                } else {
                    manifestBase
                }
                "$schemeAndHost$this"
            } else {
                "$manifestBase/$this"
            }
        }
    }

File: lib/services/addon_manifest_parser.dart
// Ported from NuvioMobile's `features/addons/AddonManifestParser.kt`.
// SPDX-License-Identifier: GPL-3.0
// Source: https://github.com/NuvioMedia/NuvioMobile (GPL-3.0)
//
// Parses a Stremio addon manifest document fetched from a transport URL.
library;

import 'dart:convert';

import 'addon_models.dart';

class AddonManifestParser {
  const AddonManifestParser();

  /// Parses [payload] (a JSON manifest document) that was fetched from
  /// [manifestUrl]. Relative logo URLs are resolved against the manifest URL.
  AddonManifest parse({
    required String manifestUrl,
    required String payload,
  }) {
    final decoded = jsonDecode(payload);
    if (decoded is! Map<String, dynamic>) {
      throw const FormatException('Add-on manifest must be a JSON object');
    }
    final root = decoded;

    final defaultTypes = _stringList(root, 'types');
    final defaultPrefixes = _stringList(root, 'idPrefixes');

    return AddonManifest(
      id: _requiredString(root, 'id'),
      name: _requiredString(root, 'name'),
      description: _optionalString(root, 'description') ?? '',
      version: _requiredString(root, 'version'),
      logoUrl: _resolveAgainstManifest(
        _optionalString(root, 'logo'),
        manifestUrl,
      ),
      resources: _resources(root, defaultTypes, defaultPrefixes),
      types: defaultTypes,
      idPrefixes: defaultPrefixes,
      catalogs: _catalogs(root),
      behaviorHints: _behaviorHints(root),
      transportUrl: manifestUrl,
    );
  }

  List<AddonResource> _resources(
    Map<String, dynamic> root,
    List<String> defaultTypes,
    List<String> defaultPrefixes,
  ) {
    final raw = root['resources'];
    if (raw is! List) return const [];
    final resources = <AddonResource>[];
    for (final element in raw) {
      if (element is String) {
        if (element.isNotEmpty) {
          resources.add(AddonResource(
            name: element,
            types: defaultTypes,
            idPrefixes: defaultPrefixes,
          ));
        }
      } else if (element is Map<String, dynamic>) {
        final name = _requiredString(element, 'name');
        resources.add(AddonResource(
          name: name,
          types: _stringList(element, 'types').isNotEmpty
              ? _stringList(element, 'types')
              : defaultTypes,
          idPrefixes: _stringList(element, 'idPrefixes').isNotEmpty
              ? _stringList(element, 'idPrefixes')
              : defaultPrefixes,
        ));
      }
    }
    return resources.where((r) => r.name.isNotEmpty).toList(growable: false);
  }

  List<AddonCatalog> _catalogs(Map<String, dynamic> root) {
    final raw = root['catalogs'];
    if (raw is! List) return const [];
    final catalogs = <AddonCatalog>[];
    for (final element in raw) {
      if (element is! Map<String, dynamic>) continue;
      final type = _requiredString(element, 'type');
      final id = _requiredString(element, 'id');
      catalogs.add(AddonCatalog(
        type: type,
        id: id,
        name: (_optionalString(element, 'name') ?? '').isEmpty
            ? id
            : _optionalString(element, 'name')!,
        extra: _extraProperties(element),
      ));
    }
    return catalogs;
  }

  List<AddonExtraProperty> _extraProperties(Map<String, dynamic> catalog) {
    final raw = catalog['extra'];
    if (raw is! List) return const [];
    final extras = <AddonExtraProperty>[];
    for (final element in raw) {
      if (element is! Map<String, dynamic>) continue;
      final name = _optionalString(element, 'name');
      if (name == null || name.isEmpty) continue;
      extras.add(AddonExtraProperty(
        name: name,
        isRequired: element['isRequired'] == true,
        options: _stringList(element, 'options'),
        optionsLimit: _optionalInt(element, 'optionsLimit'),
      ));
    }
    return extras;
  }

  AddonBehaviorHints _behaviorHints(Map<String, dynamic> root) {
    final hints = root['behaviorHints'];
    if (hints is! Map<String, dynamic>) {
      return const AddonBehaviorHints();
    }
    return AddonBehaviorHints(
      configurable: hints['configurable'] == true,
      configurationRequired: hints['configurationRequired'] == true,
      adult: hints['adult'] == true,
      p2p: hints['p2p'] == true,
    );
  }
}

String _requiredString(Map<String, dynamic> map, String name) {
  final value = _optionalString(map, name);
  if (value == null || value.isEmpty) {
    throw FormatException('Manifest is missing required field "$name"');
  }
  return value;
}

String? _optionalString(Map<String, dynamic> map, String name) {
  final value = map[name];
  if (value is String && value.isNotEmpty) return value;
  return null;
}

List<String> _stringList(Map<String, dynamic> map, String name) {
  final value = map[name];
  if (value is! List) return const [];
  return value
      .whereType<String>()
      .map((e) => e.trim())
      .where((e) => e.isNotEmpty)
      .toList(growable: false);
}

int? _optionalInt(Map<String, dynamic> map, String name) {
  final value = map[name];
  if (value is int) return value;
  if (value is String) return int.tryParse(value);
  return null;
}

/// Mirrors Nuvio's URL resolution: absolute/data URLs pass through, protocol-
/// relative URLs gain `https:`, root-relative URLs are joined to the manifest
/// origin, and other relative URLs are joined to the manifest directory.
String? _resolveAgainstManifest(String? url, String manifestUrl) {
  if (url == null || url.isEmpty) return null;
  if (url.startsWith('http://') ||
      url.startsWith('https://') ||
      url.startsWith('data:')) {
    return url;
  }
  if (url.startsWith('//')) return 'https:$url';

  final withoutQuery = manifestUrl.split('?').first;
  // Everything up to (not including) the last '/' — the manifest directory.
  final lastSlash = withoutQuery.lastIndexOf('/');
  final manifestBase = lastSlash >= 0 ? withoutQuery.substring(0, lastSlash) : '';

  if (url.startsWith('/')) {
    // Root-relative: resolve against the manifest's scheme + host.
    final schemeEnd = manifestBase.indexOf('://');
    if (schemeEnd >= 0) {
      final scheme = manifestBase.substring(0, schemeEnd + 3);
      final hostRest = manifestBase.substring(schemeEnd + 3);
      final hostEnd = hostRest.indexOf('/');
      final origin =
          '$scheme${hostEnd >= 0 ? hostRest.substring(0, hostEnd) : hostRest}';
      return '$origin$url';
    }
    return '$manifestBase$url';
  }
  return '$manifestBase/$url';
}
